clientapi/routing/redaction.go

```diff
@@ -0,0 +1,134 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	currentstateAPI "github.com/matrix-org/dendrite/currentstateserver/api"
+	"github.com/matrix-org/dendrite/internal/config"
+	"github.com/matrix-org/dendrite/internal/eventutil"
+	"github.com/matrix-org/dendrite/roomserver/api"
+	roomserverAPI "github.com/matrix-org/dendrite/roomserver/api"
+	userapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+type redactionContent struct {
+	Reason string `json:"reason"`
+}
+
+type redactionResponse struct {
+	EventID string `json:"event_id"`
+}
+
+func SendRedaction(
+	req *http.Request, device *userapi.Device, roomID, eventID string, cfg *config.Dendrite,
+	rsAPI roomserverAPI.RoomserverInternalAPI, stateAPI currentstateAPI.CurrentStateInternalAPI,
+) util.JSONResponse {
+	resErr := checkMemberInRoom(req.Context(), stateAPI, device.UserID, roomID)
+	if resErr != nil {
+		return *resErr
+	}
+
+	ev := roomserverAPI.GetEvent(req.Context(), rsAPI, eventID)
+	if ev == nil {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.NotFound("unknown event ID"), // TODO: is it ok to leak existence?
+		}
+	}
+	if ev.RoomID() != roomID {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.NotFound("cannot redact event in another room"),
+		}
+	}
+
+	// "Users may redact their own events, and any user with a power level greater than or equal
+	// to the redact power level of the room may redact events there"
+	// https://matrix.org/docs/spec/client_server/r0.6.1#put-matrix-client-r0-rooms-roomid-redact-eventid-txnid
+	allowedToRedact := ev.Sender() == device.UserID
+	if !allowedToRedact {
+		plEvent := currentstateAPI.GetEvent(req.Context(), stateAPI, roomID, gomatrixserverlib.StateKeyTuple{
+			EventType: gomatrixserverlib.MRoomPowerLevels,
+			StateKey:  "",
+		})
+		if plEvent == nil {
+			return util.JSONResponse{
+				Code: 403,
+				JSON: jsonerror.Forbidden("cannot redact event, missing pl event"),
+			}
+		}
+		pl, err := plEvent.PowerLevels()
+		if err != nil {
+			return util.JSONResponse{
+				Code: 403,
+				JSON: jsonerror.Forbidden("cannot redact event, malformed pl event"),
+			}
+		}
+		allowedToRedact = pl.UserLevel(device.UserID) >= pl.Redact
+	}
+	if !allowedToRedact {
+		return util.JSONResponse{
+			Code: 403,
+			JSON: jsonerror.Forbidden("cannot redact event"),
```

Comment:
I suspect this error gets rendered by the client - can we make it a bit more descriptive? e.g. `You don't have permission to redact this event (power level %d is required)`